File: src/main/java/net/volcano/jdacommands/interfaces/CommandClient.java
package net.volcano.jdacommands.interfaces;

import net.volcano.jdacommands.client.ReactionMenuClient;
import net.volcano.jdacommands.exceptions.command.CommandCompileException;
import net.volcano.jdacommands.model.command.Command;
import net.volcano.jdacommands.model.command.CommandCompiler;
import net.volcano.jdacommands.model.command.CommandEvent;
import net.volcano.jdacommands.model.command.CommandNode;
import net.volcano.jdacommands.model.command.arguments.interfaces.CodecRegistry;

import java.util.Set;
import java.util.concurrent.ScheduledThreadPoolExecutor;

public interface CommandClient {
	
	UserProvider getUserProvider();
	
	GuildProvider getGuildProvider();
	
	PrefixProvider getPrefixProvider();
	
	PermissionProvider getPermissionProvider();
	
	CommandCompiler getCommandCompiler();
	
	CodecRegistry getCodecRegistry();
	
	Set<Command> getAllCommands();
	
	CommandNode getRootCommandNode();
	
	ScheduledThreadPoolExecutor getExecutorService();
	
	ReactionMenuClient getReactionMenuClient();
	
	PermissionClient getPermissionClient();
	
	boolean registerCommand(Command command);
	
	boolean registerController(Object controller) throws CommandCompileException;
	
	default void call(CommandEvent event) {
		try {
			executeCommand(event);
		} catch (Exception e) {
			terminate(event, e);
		}
	}
	
	void executeCommand(CommandEvent event);
	
	void terminate(CommandEvent event, Throwable t);
	
}

File: src/main/java/net/volcano/jdacommands/interfaces/CommandClient.kt
package net.volcano.jdacommands.interfaces

import net.volcano.jdacommands.exceptions.command.CommandCompileException
import net.volcano.jdacommands.model.command.Command
import net.volcano.jdacommands.model.command.CommandCompiler
import net.volcano.jdacommands.model.command.CommandEvent
import net.volcano.jdacommands.model.command.CommandNode
import net.volcano.jdacommands.model.command.arguments.interfaces.CodecRegistry
import java.util.concurrent.ScheduledThreadPoolExecutor

interface CommandClient {

	val userProvider: UserProvider
	val guildProvider: GuildProvider
	val prefixProvider: PrefixProvider
	val permissionProvider: PermissionProvider
	val commandCompiler: CommandCompiler
	val codecRegistry: CodecRegistry
	val allCommands: Set<Command>
	val rootCommandNode: CommandNode
	val executorService: ScheduledThreadPoolExecutor
	val interactionClient: InteractionClient
	val permissionClient: PermissionClient

	fun registerCommand(command: Command): Boolean

	@Throws(CommandCompileException::class)
	fun registerController(controller: Any): Boolean

	fun call(event: CommandEvent) {
		try {
			executeCommand(event)
		} catch (e: Exception) {
			terminate(event, e)
		}
	}

	fun executeCommand(event: CommandEvent)

	fun terminate(event: CommandEvent, t: Throwable)

}